python/cudnn/sdpa/fwd/api_dsl.py

```
@@ -1236,6 +1236,36 @@ def _key(desc):
 
+    def _thd_unit_count(self, slq_host) -> int:
+        """One THD unit per CGA-height slice of each sequence's Q rows —
+        the EXACT total, host-computed from the length round-trip (legacy
+        modules only).
+
+        The kernel tolerates OVER-counting (issue #552): units past the live
+        total decode to the batch == n_batch dead sentinel — every role takes
+        the empty-KV path and neither O nor LSE is written — so a launch grid
+        sized above the exact total changes nothing but occupancy."""
+        cga_tile_m = int(self._k_mod.CGA_TILE_M)
+        return self.h_q * sum((l + cga_tile_m - 1) // cga_tile_m for l in slq_host)
```

Ruff (0.16.1)

[error] Ambiguous variable name: `l`

(E741)